harmonization of switcher styles and usage (also see #1099 style for palettechoser)

tobibeer opened the pull request:
depends on a properly refreshing set widget, see #1952

allows to override styles more easily via new class tc-chosen
# Demo

http://1954.tiddlyspot.com (updated 2016-10-24)
# Merging

If being merged, once smashed, a good overall commit message would be:

```
* hamonize all switchers to implement `tc-chooser`, higlighting the selected option via `tc-chosen`
* languageswitcher now actually used (was orphaned before)
* all respective PageControls simplified by transcluding switchers
* ui improvements for toolbar buttons in ControlPanel, full labe toggles selection
* harmonization / ui improvements for popups triggered by PageControls
* cosmetics to palettepreview tid
```

tobibeer commented:
preview

![style](https://cloud.githubusercontent.com/assets/97566/10014071/0c045914-6117-11e5-957e-5b0d169910e4.jpg)

Jermolene commented:
Need to fix this consistently with the dropdowns for storyview, theme, palette, language. Many thanks!

tobibeer commented:
updated all of storyview, theme, palette, language
* both in ControlPanel as well as PageControls
* removed "current ..." form switchers as it's redundant
* streamlined toolbar buttons in ControlPanel, especially less spacing

# Demo

http://1954.tiddlyspot.com

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -1915,20 +1921,50 @@ a.tc-tiddlylink.tc-plugin-info:hover .tc-plugin-info > .tc-plugin-info-chunk > s
+.tc-chooser-item:hover a.tc-tiddlylink:hover {
+	text-decoration: underline;
+	color: <<colour foreground>> !important;
+}
```
NOOOOOOO! important. That's toxic!

pmario replied:
This is a base CSS and needs !important. There's something really wrong!!!!

tobibeer replied:
thankx @pmario, I'll see if I can make the selector more specific instead.

tobibeer commented:
@pmario, thanks for paying attention to detail, I removed the problematic css and addressed more clearly what I wanted to override for these kinds of "chosers".

pmario commented:
did you update the tiddlyspot demo page? So I can have a look?

pmario commented:
---- OT 
I personally would really like to analyse and check TW CSS + HTML structure, using the "workflow" Nicole Sullivan talks about here. https://youtu.be/j6sAm7CLoCQ?t=355  ~20min worth viewing. 

what do you think?
-----

tobibeer commented:
> did you update the tiddlyspot demo page? So I can have a look?

Should be up to date, yes.

tobibeer commented:
polished some styles...

* chosen items now more emphasized while curspor not indicating a link
* removed underline when hovering, used highlighted border colours instead

tobibeer commented:
Did some more cosmetics and improvements on the css, I'd leave it at that for now

Please squash appropriately if merging.

Demo updated.

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -627,6 +627,10 @@ button.tc-untagged-label {
+}
+
+.tc-page-controls > span > button,
+.tc-page-controls > span > span > button{
```
look ugly.

pmario replied:
CSS rules are processed form right to left. So everytime a button is handled, the browsers CSS processor needs to check if a button is somewhere in a span which is a child somewhere in a span, which covered by any element that happens to be classed: tc-page-controls. I don't like those rules. IMO they will break soon.

pmario replied:
IMO give the button a well defined class and trash the overhead.

pmario replied:
Just made some tests. It seems that you can safely remove the whole CSS rule, because it isn't used anywhere. There is no button element in the language switcher dropdown. Have a look at the DOM.

tobibeer replied:
@pmario 

> There is no button element in the language switcher dropdown.

It's not about a button in the language switcher dropdown. It's about addressing solely the button contents in the pagecontrols, but not any popup reveals.

The point is specifically that I only want to address the buttons, not everything in PageControls and why I changed that line to begin with.

tobibeer commented:
@pmario, for comparision:

**After my changes:**

![now](https://cloud.githubusercontent.com/assets/97566/19634288/02e332d8-99ba-11e6-8ce4-49e3b5b1aa8c.jpg)

**Before my changes:**

![before](https://cloud.githubusercontent.com/assets/97566/19634296/1b5d4394-99ba-11e6-955b-77c4c2683146.jpg)

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -1792,6 +1791,12 @@ html body.tc-body.tc-single-tiddler-window {
 	width: 100%;
 }
 
+.tc-controlpanel svg.tc-image-button{
```
typo: `.tc-control-panel` 
and 
`.tc-controlpanel .tc-image-button`  ... imo has the same effect.

tobibeer replied:
@pmario, I will need to investigate what it was before. All in all, this part does not need to / should not be changed by this PR. It's an artifact.

tobibeer replied:
@pmario, going to fix that tonight, already updated the demo http://1954.tiddlyspot.com

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -1792,6 +1791,12 @@ html body.tc-body.tc-single-tiddler-window {
+.tc-controlpanel svg.tc-image-button{
+	width: 1em;
+	height: 1em;
+	vertical-align:middle;
```
vertical-align is already inhereted from:  label svg
```
button svg, button img, label svg, label img {
    vertical-align: middle;
}
```

it's the input box, that needs a vertical-alignment.

tobibeer replied:
@pmario, ah, I guess that slipped in when I was trying to figure out the right spot to put it but then didn't fully revert to the initial state

will fix this eventing

tobibeer replied:
@pmario

> it's the input box, that needs a vertical-alignment.

Which input box? This is meant to vertically align the buttons in the toolbar configuration, see updated demo http://1954.tiddlyspot.com.

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -1792,6 +1791,12 @@ html body.tc-body.tc-single-tiddler-window {
 
+.tc-controlpanel svg.tc-image-button{
+	width: 1em;
+	height: 1em;
```
set your browser zoom to 100% and imo it will be too small. so I'd suggest. 1.5em

tobibeer replied:
@pmario, I believe that was so beforehand. It's nothing I have deliberately touched, just something I experimented with to find the right spot for fixing the oversized popup contents.

We can't easily touch that (not in this context) because things like tiddler icons also depend on that, e.g. the cp cogwheel.

tobibeer replied:
@pmario, I have gone and fixed the class to `tc-control-panel` in the demo http://1954.tiddlyspot.com for review.

The buttons have the very same size as in the dropdowns, so they cannot be **too** small, since that's what the main UI is using as well.

tobibeer commented:
Will update vanilla/base again later tonight according to @pmario's comments. Although, I'm not inclined to have the button images be bigger for the toolbar, since that is what we're using in dropdowns anyways and the current preview simply is too big and impingens usability for (de-)activating toolbar buttons, imho.

(If you're wondering why I did that in the context of palettechooser then that's because I ended up reviewing half of the appearance tabs in the ControlPanel owed to harmonization wishes.)

pmario commented:
>(If you're wondering why I did that in the context of palettechooser then that's because I ended up reviewing half of the appearance tabs in the ControlPanel owed to harmonization wishes.)

I know. I think you'll like my suggestions. They are simple and generic and also work if you view the tiddlers. Not only in the ControlPanel. Even if you put the to the sidebar, should work nicely. ... And my css is much simpler.

pmario commented:
PR to your repo will follow shortly.

tobibeer commented:
@pmario, thanks for this, will merge in a bit and push an update... perhaps you may re-check later if there is some remnant css that can be removed

pmario commented:
I'm not sure, if Jeremy will merge this. IMO it may be split into 3 separated PRs. 

 1. generic checkbox input "vertical-alignment" ..... one liner in base.tid
 2. ControlPanel Toolbars stuff
 3. Theme, Palette switcher stuff.

pmario commented:
Did you have a look at the PR I sent you? https://github.com/tobibeer/TiddlyWiki5/pull/13

tobibeer commented:
The thing is, it's just not fun to make PRs that depend on the same file being modified, at all.

tobibeer commented:
merged changes from @pmario but reverted changes made to PageControl button styles

updated demo: http://1954.tiddlyspot.com

tobibeer commented:
Ok, unless @pmario has more useful recommendations I'm hoping that'll be all.

tobibeer commented:
Well, good thing I reviewed the changed files again. ;-)

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -1910,25 +1921,62 @@ a.tc-tiddlylink.tc-plugin-info:hover .tc-plugin-info > .tc-plugin-info-chunk > s
+	color: <<colour foreground>>;
+}
+
+.tc-chosen > a.tc-tiddlylink:before {
```
I think: `+.tc-chosen a.tc-tiddlylink:before {`  should also work, with much less impact on the CSS processor. 
the `a` element is allways the first:child below `tc-chosen`

pmario replied:
I did test it and it seems to work well, with the simpler rule.

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -1877,10 +1877,6 @@ a.tc-tiddlylink.tc-plugin-info:hover .tc-plugin-info > .tc-plugin-info-chunk > s
-/*
-** Checkbox tiddlers transcluded into ControlPanel. see: #1954
-*/
-
```
The other sections also have comments like those. You may want to remove #1954. but even this one is useful. imo.

tobibeer commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -1910,25 +1921,62 @@ a.tc-tiddlylink.tc-plugin-info:hover .tc-plugin-info > .tc-plugin-info-chunk > s
+.tc-chooser-item:hover {
+	background-color: <<colour table-header-background>>;
+	border-color: <<colour table-footer-background>>;
+}
```
@pmario, currently, we have no hover highlight for table borders or headers, e.g. hover states in dropdowns. The problem with using more contrasty backgrounds for the selected option in a sidebar popup is that, since link colours are grey'ish, the contrast between the text and the background will suffer if we use a darker background for the selected option. However, I deliberately did not use the inverted colours, especially with regards to the paletteswitcher.

The choosers are **not** exactly dropdowns, although used in such via PageControl popups. Those in the control center are always visible right away, and basically more like table-like lists, hence my inclination to use table styles.

When used in drop-downs, I think their "link" styles should not be overruled / fully inverted like we do with basic dropdowns, at least that doesn't work well with the language switcher and especially not the palette. Hence, my decision to not use this type of "dramatic" highlighting at all.

If tables styles end up producing very little contrast for switchers then so they do for tables. I believe, in that case, that is what the theme designer wanted.

I'm also quite fine with using the table header background for the highlighted item so as to give it a more prominent, yet subtle highlight.

In other words, I'm quite fine with the kind of gentle highlight they have atm. If you wish to play with it more, perhaps upload a demo with proposed other colors somewhere? However, I would recommend to only address dropdown related styles or to address them separately.

What I wanted to avoid was to arbitrarily pick different colours from different aspects of TiddlyWiki. So, having them all be table-based works well in that respect. But perhaps yes, the dropdown colours could be of a different origin than those used in the control center.

pmario commented:
I'm fine with the selected color, if it id low contrast.  but I'd like to have a stronger hover contrast. ... didn't  chech your latest commit yet

pmario commented:
All in all. I think this is cool stuff now. IMO being able to click the labels to activate the checkboxes is a huge usability improvement now.

tobibeer commented:
Added merging info to description:

> # Merging
> 
> If being merged, once smashed, a good overall commit message would be:
> 
> ```
> * hamonize all switchers to implement `tc-chooser`, higlighting the selected option via `tc-chosen`
> * languageswitcher now actually used (was orphaned before)
> * all respective PageControls simplified by transcluding switchers
> * ui improvements for toolbar buttons in ControlPanel, full labe toggles selection
> * harmonization / ui improvements for popups triggered by PageControls
> * cosmetics to palettepreview tid
> ```

tobibeer commented:
Tried to merge master, rebase and squash commits, but somehow it appears that changes in the demo got lost.

tobibeer commented:
Ha, of course, I need to switch to using my local repo and not the installed tiddlywiki 💃

tobibeer commented:
Ok, all done. Demo is updated to current state / release. Please test / consider merging.

twMat commented:
@Jermolene ...is this to be in next release? If not, at least give a reason, please.
